### ISSUE TITLE
Collaboration with alloc-wg and @CAD97's foundation grant

### ISSUE THREAD

**Issue by CAD97** [CraftSpider/department#6]
Oh hi! I hadn't seen this until I saw it pop up in https://github.com/rust-lang/rust/pull/97052.

I'm part of alloc-wg and about two months ago I picked up the Storage API again in this repo: https://github.com/CAD97/storages-api (docs: https://cad97.github.io/storages-api/storage_api/index.html)

As part of this refinement and simplification push, I ended up converting the `Storage` trait to be untyped: that is, `Storage::Handle` is no longer a GAT, and notably this means that handle resolution can return regular references to `[MaybeUninit<u8>]` rather than pointers with difficult to explain invalidation rules.

There's some miscellaneous renaming involved, but also the simplification brought us down to a set of just four traits: `Storage`, `SharedMutabilityStorage`, `PinningStorage`, and `MultipleStorage`. I've been advised (and agree) that controlling complexity is important here for a path to std inclusion; even with this simplified API surface without GAT, it's still a contender for the most complex API exposed by std. And that complexity is safety critical.

I'm fairly convinced that the current shape in my repo is minimal for what we want to support, but I wasn't thinking the previous POC was *terribly* over-engineered[^1] either, so it's possible that a further simplification is hiding.

[^1]: Even back then I disagreed on some details with matthieu-m, which ended up being integrated into the most recent iteration. Notably, matthieu-m had separate API trees for "`Element`" and "`Range`" storage, but I've always advocated for "`Range`" storage to just be "`Element`" storage of an array/slice. Along with this difference is that matthieu-m's element storage takes `T` to place it, whereas mine previously took `<T as Pointee>::Metadata` (when it was still typed) and now just deals in `Layout`.

matthieu-m agrees that the untyping of storage is probably a good direction; the previous storage API was not typed due to some conscious decision due to necessity, but because it evolved from C++'s `std::allocator` design, which is typed, and matthiue-m just hadn't considered using an untyped interface. This does mean that *usage* of the storage traits becomes more `unsafe`/difficult, but this is easily addressed by adding a typed `RawBox` around `Storage`, `RawVec` around `Storage` of slices, and maybe `RawArena` around `MultipleStorage`? as a bridge between raw storage and high-level collections.

(It's also quite possible that matthieu-m overlooked the possibility of making handles untyped because of the relative newness of the pointee metadata APIs at the time, in order to draw the conclusion that `S::Handle<T>` can be replaced by `(S::Handle, <T as Pointee>::Metadata)`.)

Anyway, getting to the point:

I [received a Rust Foundation Community Grant](https://foundation.rust-lang.org/news/2022-06-14-community-grants-program-awards-announcement/). My project is sort of two projects; the first is unrelated work in the tracing ecosystem, but the second half (scheduled Oct..=Dec) is to develop and push the Storage API to an MCP. https://github.com/rust-lang/rust/pull/97052 is actually an early start on prerequisite work for upstreaming storages, as otherwise a storage-generic `Box` can't be coerced.

```rust
pub struct RawBox<T: ?Sized, S: Storage> {
    handle: S::Handle,
    metadata: <T as Pointee>::Metadata, // -> JustMetadata<T>
    storage: S,
}
```

My plan for MCPing the Storage API is basically broken down into the following three parallel steps:

- Recruit a T-lang member to liason/proxy/sponsor/whatever-the-word-is the proposal.
- Provide a proof of concept implementation for minimally all currently allocator-generic std collections on top of the Storage API.
  - I'm still up-in-the-air about whether to do this in-tree on the std impls directly in a fork, or to reimplement them in a separate repo (like you've done here).
  - Part of what makes this interesting is that this makes std collections `#[no_std]` compatible, so while refactoring in-place has the benefits of full API coverage and ease of "build with my fork of std" testing, wanting to move things into libcore means the diff would be little better than an external crate. Plus having to deal with std iteration times.
- Write the MCP.
  - Document the supported use cases / design constraints / etc.
  - Document why the API has the shape it does.
    - My intent here is to guide the reader through inventing the storage traits themselves, like the "reinventing `tower::Service`" article.
  - Make the argument for putting this into std in addition to the current unstable `Allocator` trait.

Ultimately, I think that this MCP is where the Storage API either sinks or swims. Either it gets accepted for in-tree implementation/experimentation, or it gets rejected, likely on the matter of being insufficiently motivated for std inclusion given its complexity. Especially given the async-wg group w.r.t. async in traits (e.g. `dyn*`), now really does feel like the time to say if we want to make `Box` the "owning handle to some storage" or if it's just an owning pointer to allocated memory. My claim is that `Box<T, ?>` can be both `&move T` and `dyn* T`, as well as `InlineMaxSizeDst<T>` and so much more, so its a matter of proving it worth it.

**I'd like to work together**, or at the very least avoid duplicating work you've already done unnecessarily. Adding your experience working with (your modified copy of) the storage API will help to strengthen the motivation provided in the MCP.

...I don't really know how to end this info dump. I guess...

TL;DR: Hi 👋 I'm working on the Storage API and would like to incorporate your experience working with it.

**Comment by CraftSpider** [comment]
Hello! I'd love to help with pushing storages forwards towards `std` acceptance, as my biggest current pain is that I find myself wanting to just copy-paste things like `BTreeMap` or `OsString` into this repository, its ergonomics would be much better in the standard library. You can contact me through gmail at `runetynan (at) gmail (dot) com`, or through discord at `CraftSpider # 0001`, if you want to talk further.

Some quick thoughts here:

I think untyping the handle loses some ergonomics, but also understandably simplifies the API, one of my current big pains is that a typed handle can't have a useful supertrait due to lack of things like casting support.

You seem to have ended on a similar API, with some slightly different specific choices, but that may be a good point towards it as a good choice for the final API design - We separately came up with basically the same API from the original POC.

I decided instead of `SharedMutability` as an extension I'd require all storages to be internally mutable, and instead of a single `Pinnable`, I separated it into multiple traits. I think the first is fine, though for the second, I think it may actually be a useful distinction - I've found multiple extension storages that implement only a subset of my distinguished traits, and it simplifies the safety requirements and implementations for unleaking quite a bit.

**Comment by CraftSpider** [comment]
Bottom to Top:

No problem with the brain dump, I appreciate the thoughts on potential API ramifications and etc, once there's a single location these thoughts can hopefully be centralized, and probably use issues or such to track these kinds of API decisions.

The rename would *probably* be fine I think, but also that's probably the kind of thing that will get some bikeshedding later on :P

From everything you've said, I agree that bounding the handle might be nicer. My use case was more about being able to relax `S = AllocStorage<Global>`, and support leaking from stack heaps which could be useful for getting `&'storage mut T` items out of boxes. Your analysis of the traits seems right, you skipped over `ClonesafeStorage`, but it's basically just a finer-grained refinement on `LeaksafeStorage`, and probably not necessary.
